add CRF module to knlp

ERICMIAO0817 opened the pull request:
重新调整了pr，并且把inference也封装成类便于后续拓展。

ERICMIAO0817 commented:
新增加拼音输入法模块

DukeEnglish commented on knlp/seq_labeling/pinyin_input_method/inference.py:
```diff
@@ -0,0 +1,195 @@
+import math
+from knlp.seq_labeling.crf.crf import CRFModel
+
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__))) + "/../.."
```
修改了路径之后，这里也不能统一使用定义的base么？

DukeEnglish commented on knlp/seq_labeling/crf/crf.py:
```diff
@@ -0,0 +1,106 @@
+from knlp.seq_labeling.crf.crf_utils import sentence2features
+from sklearn_crfsuite import CRF
+
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__))) + "/../.."
```
这个记得修改一下

DukeEnglish commented on knlp/seq_labeling/crf/crf_utils.py:
```diff
@@ -0,0 +1,25 @@
+# -*-coding:utf-8-*-
+# ******** CRF 工具函数*************
+
+def word2features(sent, i):
```
在文档中，可以标明 这个函数是可以修改的。

DukeEnglish commented on knlp/seq_labeling/crf/inference.py:
```diff
@@ -0,0 +1,81 @@
+
+        re_zh, re_no_zh = re.compile("([\u4E00-\u9FA5]+)"), re.compile("[^a-zA-Z0-9+#\n]")  # 只对汉字做分词
+        processed_sentence = re_zh.split(in_put)  # 按照汉字团进行分割
+        # out_sent = []
```
注释没有用就删除

DukeEnglish commented on knlp/seq_labeling/crf/inference.py:
```diff
@@ -0,0 +1,81 @@
+            if re_zh.match(block):  # 对汉字进行分词
+                blocking = list(block)
+                pred = [blocking]
+                print(pred)
```
这个没有用就删除掉吧

DukeEnglish commented on knlp/seq_labeling/crf/inference.py:
```diff
@@ -0,0 +1,81 @@
+        """
+        crf = CRFModel()
+
+        re_zh, re_no_zh = re.compile("([\u4E00-\u9FA5]+)"), re.compile("[^a-zA-Z0-9+#\n]")  # 只对汉字做分词
```
这里有一个小细节，你的这个操作会不会导致标点符号被删除？尝试输入一下 带有标点的句子，看看输出结果和预期是否一致

DukeEnglish commented on knlp/seq_labeling/crf/train.py:
```diff
@@ -1,9 +1,63 @@
+
+        self.model.train(words, tags)
+
+    def save_model(self, file_name):
```
注释说明 filename是输出路径

DukeEnglish commented on knlp/seq_labeling/crf/train.py:
```diff
@@ -1,9 +1,63 @@
+    """
+
+    def __init__(self, data_path):
+
```
注释说明这个datapath是什么

DukeEnglish commented on knlp/seq_labeling/pinyin_input_method/pinyin_segment_train.py:
```diff
@@ -0,0 +1,63 @@
+
+    def __init__(self, data_path):
+
+        self.training_data_path = data_path
```
添加注释

DukeEnglish reviewed (commented):
有一些小细节，我备注了，修改一下。本周末修改好告诉我一声，没什么问题就提交了

ERICMIAO0817 commented:
收到，没有问题！




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "global-nlp/knlp"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年3月26日(星期六) 晚上6:42
***@***.***&gt;;
抄送:&nbsp;"Captain ***@***.******@***.***&gt;;
主题:&nbsp;Re: [global-nlp/knlp] add CRF module to knlp (PR #5)





 
@DukeEnglish commented on this pull request.
 
有一些小细节，我备注了，修改一下。本周末修改好告诉我一声，没什么问题就提交了

 
In knlp/seq_labeling/pinyin_input_method/inference.py:
 &gt; @@ -0,0 +1,195 @@ +# coding: utf-8 +import os +import json +import math +from knlp.seq_labeling.crf.crf import CRFModel + +BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__))) + "/../.."  
修改了路径之后，这里也不能统一使用定义的base么？
 
 
In knlp/seq_labeling/crf/crf.py:
 &gt; @@ -0,0 +1,106 @@ +# -*-coding:utf-8-*- +""" +针对CRF模型的类定义 +（注：导入pickle用于储存对象化的模型，便于存储用于不同任务的训练后的模型。） + +CRFModel类： +由sklearn_crfsuite中的CRF类初始化而来。其中包括训练算法、l1正则与l2正则的系数，迭代次数等模型参数。 +""" +import os +import pickle + +from knlp.seq_labeling.crf.crf_utils import sentence2features +from sklearn_crfsuite import CRF + +BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__))) + "/../.."  
这个记得修改一下
 
 
In knlp/seq_labeling/crf/crf_utils.py:
 &gt; @@ -0,0 +1,25 @@ +# -*-coding:utf-8-*- +# ******** CRF 工具函数************* + +def word2features(sent, i):  
在文档中，可以标明 这个函数是可以修改的。
 
 
In knlp/seq_labeling/crf/inference.py:
 &gt; +class Inference: + +    def __init__(self): + +        self.label_prediction = []  # 预测标签序列 +        self.out_sentence = []  # 预测结果序列 + +    def spilt_predict(self, in_put, file_path): +        """ +        将输入序列分割为各个汉字团，依次送入预训练模型中，返回各个汉字团的预测结果。 +        """ +        crf = CRFModel() + +        re_zh, re_no_zh = re.compile("([\u4E00-\u9FA5]+)"), re.compile("[^a-zA-Z0-9+#\n]")  # 只对汉字做分词 +        processed_sentence = re_zh.split(in_put)  # 按照汉字团进行分割 +        # out_sent = []  
注释没有用就删除
 
 
In knlp/seq_labeling/crf/inference.py:
 &gt; +        """ +        将输入序列分割为各个汉字团，依次送入预训练模型中，返回各个汉字团的预测结果。 +        """ +        crf = CRFModel() + +        re_zh, re_no_zh = re.compile("([\u4E00-\u9FA5]+)"), re.compile("[^a-zA-Z0-9+#\n]")  # 只对汉字做分词 +        processed_sentence = re_zh.split(in_put)  # 按照汉字团进行分割 +        # out_sent = [] +        # label_prediction = [] +        crf_model = crf.load_model(file_path) + +        for block in processed_sentence: +            if re_zh.match(block):  # 对汉字进行分词 +                blocking = list(block) +                pred = [blocking] +                print(pred)  
这个没有用就删除掉吧
 
 
In knlp/seq_labeling/crf/inference.py:
 &gt; + + +class Inference: + +    def __init__(self): + +        self.label_prediction = []  # 预测标签序列 +        self.out_sentence = []  # 预测结果序列 + +    def spilt_predict(self, in_put, file_path): +        """ +        将输入序列分割为各个汉字团，依次送入预训练模型中，返回各个汉字团的预测结果。 +        """ +        crf = CRFModel() + +        re_zh, re_no_zh = re.compile("([\u4E00-\u9FA5]+)"), re.compile("[^a-zA-Z0-9+#\n]")  # 只对汉字做分词  
这里有一个小细节，你的这个操作会不会导致标点符号被删除？尝试输入一下 带有标点的句子，看看输出结果和预期是否一致
 
 
In knlp/seq_labeling/crf/train.py:
 &gt; +            word_list = [] +            tag_list = [] +            for line in f: +                if line != '\n': +                    word, tag = line.strip('\n').split() +                    word_list.append(word) +                    tag_list.append(tag) +                else: +                    words.append(word_list) +                    tags.append(tag_list) +                    word_list = [] +                    tag_list = [] + +        self.model.train(words, tags) + +    def save_model(self, file_name):  
注释说明 filename是输出路径
 
 
In knlp/seq_labeling/crf/train.py:
 &gt; +# -*-coding:utf-8-*- +import os +import pickle + +from knlp.seq_labeling.crf.crf import CRFModel + +BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__))) + "/../.." + + +class Train: +    """ +    这个类要实现针对不同task的训练语料数据的加载，构建并保存对应模型。 +    """ + +    def __init__(self, data_path): +  
注释说明这个datapath是什么
 
 
In knlp/seq_labeling/pinyin_input_method/pinyin_segment_train.py:
 &gt; +import os +import pickle + +from knlp.seq_labeling.crf.crf import CRFModel + +BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__))) + "/../.." + + +class Train: +    """ +    这个类要实现针对不同task的训练语料数据的加载，构建并保存对应模型。 +    """ + +    def __init__(self, data_path): + +        self.training_data_path = data_path  
添加注释
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;
